crates/lib/src/compiler/rpcq.rs

```diff
@@ -76,27 +106,66 @@ impl Client {
 
     /// Retrieve the raw bytes of a response
     pub(crate) fn receive_raw(&self) -> Result<Vec<u8>, Error> {
-        self.socket.recv_bytes(0).map_err(Error::Communication)
+        self.socket
+            .lock()
+            .map_err(|e| Error::ZmqSocketLock(e.to_string()))?
+            .recv_bytes(0)
+            .map_err(Error::Communication)
+    }
+}
+
+impl quilc::Client for Client {
+    #[cfg_attr(feature = "tracing", tracing::instrument(level = "trace"))]
+    fn compile_program(
+        &self,
+        quil: &str,
+        isa: quilc::TargetDevice,
+        options: quilc::CompilerOpts,
+    ) -> Result<quilc::CompilationResult, quilc::Error> {
+        #[cfg(feature = "tracing")]
+        tracing::debug!(compiler_options=?options, "compiling quil program with quilc (RPCQ)",);
+        let params = quilc::QuilcParams::new(quil, isa).with_protoquil(options.protoquil);
+        let request = RPCRequest::new("quil_to_native_quil", &params).with_timeout(options.timeout);
+        match self.run_request::<_, quilc::QuilToNativeQuilResponse>(&request) {
+            Ok(response) => Ok(quilc::CompilationResult {
+                program: Program::from_str(&response.quil).map_err(quilc::Error::Parse)?,
+                native_quil_metadata: response.metadata,
+            }),
+            Err(source) => Err(quilc::Error::from_quilc_error(
+                self.endpoint.clone(),
+                source,
+            )),
+        }
     }
 }
 
 /// All of the possible errors for this module
 #[derive(Debug, thiserror::Error)]
 pub enum Error {
+    /// The ZMQ socket could not be created
     #[error("Could not create a socket.")]
     SocketCreation(#[source] zmq::Error),
+    /// Failed to set up auth for ZMQ
     #[error("Failed while trying to set up auth. This is likely a bug in this library.")]
     AuthSetup(#[source] zmq::Error),
+    /// Encountered error when communicating with server
     #[error("Trouble communicating with the ZMQ server")]
     Communication(#[source] zmq::Error),
+    /// Failed to serialize request
     #[error("Could not serialize request as MessagePack. This is a bug in this library.")]
     Serialization(#[from] rmp_serde::encode::Error),
+    /// Failed to deserialize response
     #[error("Could not decode ZMQ server's response. This is likely a bug in this library.")]
     Deserialization(#[from] rmp_serde::decode::Error),
+    /// Response ID did not match request ID
     #[error("Response ID did not match request ID")]
     ResponseIdMismatch,
+    /// Server responded with an error message
     #[error("Received error message from server: {0}")]
     Response(String),
+    /// Error occurred when trying to lock the ZMQ socket
+    #[error("Could not lock RPCQ client: {0}")]
+    ZmqSocketLock(String),
```

Comment:
Could this be `#[from] zmq::WhateverTheZmqSocketLockErrorIs`?

Reply:
It is possible. The error type is `PoisonError<Guard<'a, Socket>` and thus requires introducing lifetimes wherever the `Error` enum is used. Not fundamentally opposed to that... it just seemed heavy-handed.

Reply:
Ah yeah, that doesn't seem worth it to preserve the error type.